cli_tools/google-osconfig-agent/lock_windows.go

```diff
@@ -0,0 +1,88 @@
+//  Copyright 2018 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+	"unsafe"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/google-osconfig-agent/logger"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	kernel32         = windows.NewLazySystemDLL("kernel32.dll")
+	procLockFileEx   = kernel32.NewProc("LockFileEx")
+	procUnlockFileEx = kernel32.NewProc("UnlockFileEx")
+)
+
+const (
+	// https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-lockfileex
+	LOCKFILE_EXCLUSIVE_LOCK   = 2
+	LOCKFILE_FAIL_IMMEDIATELY = 1
+)
+
+func lockFileEx(hFile uintptr, dwFlags, nNumberOfBytesToLockLow, nNumberOfBytesToLockHigh uint32, lpOverlapped *syscall.Overlapped) (err error) {
+	ret, _, _ := procLockFileEx.Call(
+		hFile,
+		uintptr(dwFlags),
+		0,
+		uintptr(nNumberOfBytesToLockLow),
+		uintptr(nNumberOfBytesToLockHigh),
+		uintptr(unsafe.Pointer(lpOverlapped)),
+	)
+	// If the function succeeds, the return value is nonzero.
+	if ret == 0 {
+		return fmt.Errorf("nonzero return code from LockFileEx: %d", ret)
+	}
+	return nil
+}
+
+func unlockFileEx(hFile uintptr, nNumberOfBytesToLockLow, nNumberOfBytesToLockHigh uint32, lpOverlapped *syscall.Overlapped) (err error) {
+	ret, _, _ := procUnlockFileEx.Call(
+		hFile,
+		0,
+		uintptr(nNumberOfBytesToLockLow),
+		uintptr(nNumberOfBytesToLockHigh),
+		uintptr(unsafe.Pointer(lpOverlapped)),
+	)
+	// If the function succeeds, the return value is nonzero.
+	if ret == 0 {
+		return fmt.Errorf("nonzero return code from UnlockFileEx: %d", ret)
```

Comment:
Is `0` success or fail? It looks like this will only happen when `ret == 0` which contradicts the error message.

Reply:
fixed